### PR TITLE
Add support for connecting to signals with details

### PR DESCRIPTION
**Pull request by zec**
Some GObject signals use a non-default [_detail_ argument](https://developer.gnome.org/gobject/stable/signal.html#signal-detail). This allows one to specify that a closure should only be invoked on a subset of that signal's emissions.

There is an attribute in the GIR annotations that specifies when a signal uses non-default values of _detail_. As of time of writing, the code generated by the `gir` tool does not allow one to specify the value of _detail_ in the relevant connect methods&mdash;the closure passed to `connect_<signal name>` will get invoked on all emissions of that signal, regardless of the value of _detail_. This PR contains an attempt at adding such support in. For signals that use _detail_, the relevant method's signature is changed from, e.g.,

```rust
fn connect_state_change<F: Fn(&Self, &str, bool) + 'static>(
    &self,
    f: F
) -> SignalHandlerId
```
to

```rust
fn connect_state_change<F: Fn(&Self, &str, bool) + 'static>(
    &self,
    detail: Option<&str>,
    f: F
) -> SignalHandlerId
```

`Option<&str>` is used because not specifying a _detail_ value in `g_signal_connect*` is valid in these cases (it means the closure will be invoked regardless of the value of _detail_).

This has been tested out on code generation for gtk-rs/gtk-rs and gtk-rs/webkit2gtk-rs, and it produces the expected changes. I have even tested out the use of `Some(_)` and `None` for such a modified connection method (`webkit2gtk::UserContentManagerExt::connect_script_message_received`) and it appears to work as desired.

Some possible issues with my PR as it stands:

- This makes backward-incompatible changes to the API for these signals. Is that OK? If not, what approach(es) would be preferable?
- Is `Option<&str>` the best type to specify the _detail_ value with, or would a bespoke enum serve better?
- This is my first time working in this codebase. I wouldn't mind a sanity check of my work.

Constructive critique of this PR is appreciated!

### CODE REVIEW

**Comment by sdroege** [comment]
Thanks for the change, I think that's a useful addition.

> This makes backward-incompatible changes to the API for these signals. Is that OK? If not, what approach(es) would be preferable?

That's generally ok, yes.

> Is `Option<&str>` the best type to specify the _detail_ value with, or would a bespoke enum serve better?

What kind of enum are you thinking of?

----

To make reviewing this and understanding the impact a bit easier, can you also submit a PR to `gtk-rs/gtk-rs` that regenerates everything with this version of gir? Thanks :)

File: src/writer/to_code.rs
Comment (MarijnS95) [inline src/writer/to_code.rs:140]:
```suggestion
                        r#"let signal_name_ptr = detailed_signal_name.as_ref().map_or(&b"{0}\0"[..], |name| name.as_bytes());"#,
```

Something like that should be a bit more idiomatic in Rust. Then move the `as_ptr()` into the function call to loose the lifetime as late as possible.

File: src/writer/to_code.rs
Comment (MarijnS95) [inline src/writer/to_code.rs:151]: Perhaps you could just merge this with the `if` block above, and add an `else` for it?

```suggestion
                    v.push("connect_raw(self.as_ptr() as *mut _, signal_name_ptr.as_ptr() as *const _,".to_string());
                } else {
                    v.push(format!(
                        "connect_raw(self.as_ptr() as *mut _, b\"{}\\0\".as_ptr() as *const _,",
                        signal
                    ));
                }
```

**Comment by zec** [comment]
> > Is `Option<&str>` the best type to specify the _detail_ value with, or would a bespoke enum serve better?
> 
> What kind of enum are you thinking of?

Well, what I'm wondering there is that the variant name `None` might be a little misleading in this context. When you pass in `None` as the `detail` argument, the semantics are actually "get invoked for _any_ value of _detail_", not "get invoked for _no_ value of _detail_". As an alternative, I was thinking something along the lines of:

```rust
pub enum DetailSpecification_ExactNamesTBD {
    /// Get invoked when the detail argument matches the `str`
    Specific(&str),
    /// Get invoked for any value of the detail argument
    Any
}
```

The advantage of a new enum is that the variant names match the semantics of this use-case. The advantage of using `Option<&str>` is that it's familiar to everyone who's written non-trivial Rust.

> To make reviewing this and understanding the impact a bit easier, can you also submit a PR to `gtk-rs/gtk-rs` that regenerates everything with this version of gir? Thanks :)

I'll try to make that today (UTC-5).

File: src/writer/to_code.rs
Comment (zec) [reply src/writer/to_code.rs:151]: That makes sense. I'll unify those two conditionals.

File: src/writer/to_code.rs
Comment (zec) [reply src/writer/to_code.rs:140]: Heh, at one point in my development I had something similar at that point, then wondered if it was too convoluted. I'll make something like that change, and clarify the lifetime of `detailed_signal_name` as well.

File: src/writer/to_code.rs
Comment (MarijnS95) [reply src/writer/to_code.rs]: :+1: Sorry for CI noise, this diff already included the move of `.as_ptr()` to the call site, which requires the other diff to be applied as well.

Comment (zec) [reply src/writer/to_code.rs]: No prob, will make sure all's good in the next commit.

File: src/writer/to_code.rs
Comment (MarijnS95) [reply src/writer/to_code.rs:140]: That's fair. I always prefer `map_or` over `if let` as it has a nice default case, but always get bitten by the missing `.as_ref()` to not move out of the `Option`.

I don't think it's necessary to "clarify" the lifetime, just as long as `signal_name_ptr` is _not_ a raw pointer. That way it carries the lifetime of `detailed_signal_name` during the remainder of the function scope, disallowing that to be moved or mutated.

Comment (zec) [reply src/writer/to_code.rs]: See what you think with commit 65cb60e50efa.

Comment (MarijnS95) [reply src/writer/to_code.rs]: You don't need the explicit drop at the end, `detailed_signal_name` will go out of scope anyway. Rather, my suggestion to solve it this way (by making `signal_name` a `&[u8]` borrow) is to disallow `detailed_signal_name` from being mutably borrowed or moved at the same time. By the way, you don't need to explicitly specify the `&[u8]` type, as shown in the suggestion above.

Comment (zec) [reply src/writer/to_code.rs]: My thought with the drop is that it makes it explicit that `detailed_signal_name` outlasts the call to `connect_raw()`, but if you don't think it needs to be explicit, I can change it back.

Comment (MarijnS95) [reply src/writer/to_code.rs]: It doesn't need to be explicit, and by placing `.as_ptr()` inside the call expression (rather than before it) that is already guaranteed.

Comment (zec) [reply src/writer/to_code.rs]: OK, changed in 3d8dc87.

Comment (MarijnS95) [reply src/writer/to_code.rs]: Cool, thanks! I'm not a maintainer (can't approve this PR in a sensible way), but they'll likely ask you to squash the patches.

Comment (sdroege) [reply src/writer/to_code.rs]: Yes, please clean up the commit history a bit :)

Comment (zec) [reply src/writer/to_code.rs]: OK, I'll do that once everyone's done reviewing changes.

**Comment by sdroege** [comment]
@bilelmoussaoui Does this do the right thing on `gtk4`?

**Review by sdroege — APPROVED** [review]

**Comment by sdroege** [comment]
@MarijnS95 Thanks for the review, there was nothing left for me to complain about :)

**Review by bilelmoussaoui — APPROVED** [review]

**Comment by bilelmoussaoui** [comment]
> @bilelmoussaoui Does this do the right thing on `gtk4`?

Seems to work fine

**Comment by bilelmoussaoui** [comment]
I wonder why the `ObjectExt:: emit_with_details` takes the details as `glib::Quark` while we use a `str` here, does it make sense?

**Comment by sdroege** [comment]
> I wonder why the `ObjectExt:: emit_with_details` takes the details as `glib::Quark` while we use a `str` here, does it make sense?

`emit_by_name()` (the string version) can also take the detail as part of the string (`"signal_name::detail"`). `emit_with_details()` works with the signal id and quark. Usually if you care about using the signal id for performance reasons, you would also care about caching the quark for the detail string for multiple calls.

Using a quark here in this PR would make it harder for the majority of callers.

**Comment by sdroege** [comment]
Waiting for @GuillaumeGomez and then let's get this in :)

**Review by MarijnS95 — APPROVED** [review]

File: src/codegen/signal_body.rs
Comment (GuillaumeGomez) [inline src/codegen/signal_body.rs:31]: That's a weird name for a setter... What about `detailed` or something along the line?

**Review by GuillaumeGomez — APPROVED** [review]
Just a nit for me but otherwise looks good!

**Comment by zec** [comment]
OK, since everyone's approved the PR as it currently stands (unless I'm misinterpreting @GuillaumeGomez's comments), I've squashed the history.

**Comment by GuillaumeGomez** [comment]
It's fine if merged as is, which is why I approved it. I'd just prefer a better naming for the method though but you can ignore me. :p

**Pull request merged**
